patches/server/0889-Allow-change-origin-for-EnderDragon.patch

```diff
@@ -0,0 +1,80 @@
+     private Explosion explosionSource = new Explosion(null, this, null, null, Double.NaN, Double.NaN, Double.NaN, Float.NaN, true, Explosion.BlockInteraction.DESTROY); // CraftBukkit - reusable source for CraftTNTPrimed.getSource()
++    // Paper start - add var for save custom origin
++    @Nullable
++    private BlockPos origin = null;
++    // Paper end
+ 
+     public EnderDragon(EntityType<? extends EnderDragon> entitytypes, Level world) {
+         super(EntityType.ENDER_DRAGON, world);
+@@ -120,6 +124,19 @@ public class EnderDragon extends Mob implements Enemy {
+         return Mob.createMobAttributes().add(Attributes.MAX_HEALTH, 200.0D);
+     }
+ 
++    // Paper start
++    public BlockPos getOrigin() {
++        if (this.origin == null) {
++            return this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, EndPodiumFeature.END_PODIUM_LOCATION);
++        }
++        return this.origin;
++    }
++
++    public void setOrigin(BlockPos blockPos) {
++        this.origin = blockPos;
++    }
++    // Paper end
++
+     @Override
+     public boolean isFlapping() {
+         float f = Mth.cos(this.flapTime * 6.2831855F);
+@@ -947,7 +964,7 @@ public class EnderDragon extends Mob implements Enemy {
+                 d0 = segment2[1] - segment1[1];
+             }
+         } else {
+-            BlockPos blockposition = this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, EndPodiumFeature.END_PODIUM_LOCATION);
++            BlockPos blockposition = this.getOrigin(); // Paper - use custom origin
+             double d1 = Math.max(Math.sqrt(blockposition.distToCenterSqr(this.position())) / 4.0D, 1.0D);
+ 
+             d0 = (double) segmentOffset / d1;
+@@ -974,7 +991,7 @@ public class EnderDragon extends Mob implements Enemy {
+                 vec3d = this.getViewVector(tickDelta);
+             }
+         } else {
+-            BlockPos blockposition = this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, EndPodiumFeature.END_PODIUM_LOCATION);
++            BlockPos blockposition = this.getOrigin(); // Paper - use custom origin
+ 
+             f1 = Math.max((float) Math.sqrt(blockposition.distToCenterSqr(this.position())) / 4.0F, 1.0F);
+             float f3 = 6.0F / f1;
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragon.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragon.java
+index cd487bd68f9a8177ae3e015b3a5d1bc469743f48..42b6da996627b436bdda1823cb202ba9a0412ecb 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragon.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragon.java
+@@ -79,4 +79,16 @@ public class CraftEnderDragon extends CraftMob implements EnderDragon {
+     public int getDeathAnimationTicks() {
+         return this.getHandle().dragonDeathTime;
+     }
++
++    // Paper start
++    @Override
++    public org.bukkit.Location getOrigin() {
++        return null;
```

Comment:
You need to call the NMS methods in here.

Reply:
i keep this for the final and forget add when commit, xd

Added logic